core/silkworm/consensus/clique/clique_snapshot.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+   Copyright 2021 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#ifndef SILKWORM_CONSENSUS_CLIQUE_SNAPSHOT
+#define SILKWORM_CONSENSUS_CLIQUE_SNAPSHOT
+
+#include <map>
+#include <silkworm/common/base.hpp>
+#include <nlohmann/json.hpp>
+#include <silkworm/common/util.hpp>
+#include <silkworm/types/block.hpp>
+#include <silkworm/consensus/validation.hpp>
+#include <deque>
+
+namespace silkworm {
+
+extern std::array<uint8_t, 8> kNonceAuthorize;
+extern std::array<uint8_t, 8> kNonceUnauthorize;
+
+constexpr uint64_t kDiffInTurn = 2; // Block difficulty for in-turn signatures
+constexpr uint64_t kDiffNoTurn = 1; // Block difficulty for out-of-turn signatures
+// CliqueConfig is the consensus engine configs for proof-of-authority based sealing.
+struct CliqueConfig {
+	uint64_t   period; // Number of seconds between blocks to enforce
+    uint64_t   epoch;  // Epoch length to reset votes and checkpoint
+};
+
+// Tally is a simple vote tally to keep the current score of votes. Votes that
+// go against the proposal aren't counted since it's equivalent to not voting.
+struct Tally {
+	bool     authorize;                // Whether the vote is about authorizing or kicking someone
+	uint64_t votes;                    // Number of votes until now wanting to pass the proposal
+    std::vector<evmc::address> voters; // Voters for current proposal
+};
+
+// Voting Snapshot for Clique
+class CliqueSnapshot {
+    public:
+        CliqueSnapshot() = default;
+        CliqueSnapshot(uint64_t block_number, evmc::bytes32 hash, std::vector<evmc::address> signers,
+                       std::deque<evmc::address> recents):
+                        block_number_{block_number}, hash_{hash}, signers_{signers}, recents_{recents} {}
+
+        //! \brief Updated snapshot by adding headers
+        //! \param headers: list of headers to add.
+        //! \param config: clique config.
+        ValidationResult add_header(const BlockHeader& header, const evmc::address& signer, const CliqueConfig& config);
+        //! \brief Verify seal for header
+        //! \param header: header to verify.
+        ValidationResult verify_seal(BlockHeader header, const evmc::address& signer);
```

Comment:
`header` should be passed by `const &` (`BlockHeader` is not so small, it's expensive to pass it by value). Also, `verify_seal()` should be marked `const` since it's a read-only method.